### ISSUE TITLE
No connection could be made because the target machine actively refused it

### ISSUE THREAD

**Issue by PRGSimon** [jstedfast/MailKit#545]
I was trying to connect to the Outlook imap-server, but I got some weird exception I couldnt explain.

> System.Net.Sockets.SocketException
>   HResult=0x80004005
>   Message = No connection could be made because the target machine actively refused it
>   Source = System
>   StackTrace:
>    System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
>    System.Net.Sockets.Socket.Connect(EndPoint remoteEP)
>    System.Net.Sockets.Socket.Connect(IPAddress address, Int32 port)
>    MailKit.Net.Imap.ImapClient.Connect(String host, Int32 port, SecureSocketOptions options, CancellationToken cancellationToken)
>    MailKit.MailService.Connect(String host, Int32 port, Boolean useSsl, CancellationToken cancellationToken)
>    MailBotGui.MB.Monitor_inbox() in C:\Users\USER\documents\visual studio 2017\Projects\MB\MailBotGui\MainForm.cs: Line126
>    System.Threading.ThreadHelper.ThreadStart_Context(Object state)
>    System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
>    System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
>    System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
>    System.Threading.ThreadHelper.ThreadStart()

Here is the code where the exception showed up:

```
using (var client = new ImapClient(new ProtocolLogger("imap.log")))
            {
                // For demo-purposes, accept all SSL certificates
                client.ServerCertificateValidationCallback = (s, c, h, e) => true;

                client.Connect("imap-mail.outlook.com", 993, true);

                // Note: since we don't have an OAuth2 token, disable
                // the XOAUTH2 authentication mechanism.
                client.AuthenticationMechanisms.Remove("XOAUTH2");

                client.Authenticate("MyUsername", "MyPassword");

                // The Inbox folder is always available on all IMAP servers...
                var inbox = client.Inbox;
                inbox.Open(FolderAccess.ReadOnly);

                MessageBox.Show("Total messages: " + inbox.Count);
                MessageBox.Show("Recent messages: " + inbox.Recent);

                for (int i = 0; i < inbox.Count; i++)
                {
                    var message = inbox.GetMessage(i);
                    MessageBox.Show("Subject: " + message.Subject);
                }

                client.Disconnect(true);
            }
```

This is the line where the exception was thrown:

`client.Connect("imap-mail.outlook.com", 993, true);`

For testing purposes I used the example code.
The log file that was created contains nothing. It is completly empty.

I appreciate any help!

Thanks
Simon

**Comment by jstedfast** [comment]
The problem is that the server is not accepting connections. That's pretty much the gist of it.

Nothing can be done about it from the client side.

**Comment by abhiquiet** [comment]
Hello every one, 
I am trying to connect the pop3 port. 
I am able to telnet the server on the port, and also able to connect using the other client application, but when I try to connect using the code, I am getting the same error. 
**_No connection could be made because the target machine actively refused it [::ffff:10.123.2.123]:110_**
IPs changed here in the message
I do not know, why is it showing "::ffff" before the server IP.
kindly let me know, if I am going wrong somewhere.

Abhinav Aggarwal

**Comment by jstedfast** [comment]
What is happening is that `Socket.Connect(string host, int port)` is choosing to use the IPv6 IP address instead of the IPv4 address. Most likely your machine has an IPv6 and an IPv4 ethernet interfaces and Socket.Connect() is choosing the wrong one.

Could you try installing the MailKit package from MyGet (link is on MailKit's main github page toward the top in the Downloads table). What I did a week or 2 ago was to rewrite the connect logic to use `Socket.Connect (IPAddress address, int port)` instead and to try each resolved IP address in sequence until one connects.

**Comment by CirrusShield** [comment]
Hello Jeffrey, same issue here, we downloaded the latest version (MailKit 2.1.3.17) from MyGet and it solved the problem.

Thanks!!

**Comment by jstedfast** [comment]
Awesome, thanks for confirming the fix! I'll try to make time to push an official release later this week (or this weekend).

**Comment by hizick** [comment]
hi, I'm still having this issue. I am using the latest version of mailkit.
Thanks.

**Comment by NatriumDev** [comment]
I'm having this issue with 3.4.0
Downgrading tot 3.3.0 solved the issue for me.

**Comment by jstedfast** [comment]
Ah, I went back to `Socket.Connect (string host, int port)` in 3.4.0 to simplify code. Forgot I ever tried this before and it didn't work.

Still, you'd think that this should work.

You should file a bug against the dotnet runtime so that they fix this. It's clearly a bug on their side.

**Comment by ekalchev** [comment]
@jstedfast do you plan to revert this change? I remember we had issues with that in the past.

**Comment by jstedfast** [comment]
Yes, I plan to revert it. I'm also planning a 3.4.1 release to coincide with a 3.4.1 MimeKit release that will be getting some fixes that I am working on for Microsoft (I'm actually getting a chance to work on MimeKit for Microsoft for a few days for once!)

**Comment by jstedfast** [comment]
Reverted the change and just released MailKit v3.4.1 with the fix.
